### PR TITLE
Shallow scanning of specific paths

### PR DESCRIPTION
**Pull request by steffenclemens**
Adding feature requested in #1168.
Close https://github.com/nextcloud/groupfolders/issues/1168

### CODE REVIEW

**Comment by solracsf** [comment]
Thanks. Can you fix DCO?

**Comment by zak39** [comment]
Hi @steffenclemens :slightly_smiling_face: 

To fix the DCO, you can follow these steps :point_right: https://github.com/nextcloud/groupfolders/pull/2232/checks?check_run_id=10651495499

**Comment by steffenclemens** [comment]
> Hi @steffenclemens 🙂
> 
> To fix the DCO, you can follow these steps 👉 https://github.com/nextcloud/groupfolders/pull/2232/checks?check_run_id=10651495499

Hi @zak39, thanks for the hint.

**Comment by solracsf** [comment]
No, wait for reviews and merge. ⏳

**Comment by steffenclemens** [comment]
> No, wait for reviews and merge. ⏳

Alright, thanks! 🙂

File: lib/Command/Scan.php [hunk 1]
```diff
@@ -46,6 +46,17 @@ protected function configure() {
 				null,
 				InputOption::VALUE_NONE,
 				'Scan all the group folders.'
+			)
+			->addOption(
+				'path',
+				'p',
+				InputArgument::OPTIONAL,
```
Comment (icewind1991) [inline lib/Command/Scan.php:53]: This should be `InputOption::VALUE_REQUIRED`.

**Review by icewind1991 — DISMISSED** [review]
looks good otherwise

File: lib/Command/Scan.php (hunk 1 above)
Comment (steffenclemens) [reply lib/Command/Scan.php:53]: Hi @icewind1991, thanks for your review. Can you explain, why this should be required, if it wasn't before (as it was not existing). If it is not set, we fall back to scan the root of the group folder. In my opinion it is better for backward compatibility to keep it optional, but I wouldn't be against if you argue for "required".

Comment (steffenclemens) [reply lib/Command/Scan.php:53]: I just checked, that for the scan command of the files app, it is the same behaviour and also optional (when omitted also scanning from root). https://github.com/nextcloud/server/blob/master/apps/files/lib/Command/Scan.php#L92

Comment (icewind1991) [reply lib/Command/Scan.php:53]: (bit late responding here :see_no_evil: )

the "required" part is about value of the option, not the option itself (you can pass just `--path` as option, you have to do `--path /foo`).

`InputArgument::OPTIONAL` just happens to be the same value as `InputOption::VALUE_REQUIRED` by chance so the file scan command works as expected by accident (and should be adjusted).

Comment (silanea) [reply lib/Command/Scan.php:53]: Is there anything one could do to help move this issue forward?

**Comment by Ornanovitch** [comment]
Sorry to bump this but is there any hope to see this implemented soon? This would be a really great improvement..

**Comment by steffenclemens** [comment]
> Sorry to bump this but is there any hope to see this implemented soon? This would be a really great improvement..

Thanks for the reminder! I added the required commit.

**Comment by Ornanovitch** [comment]
Thanks :)

Will it work with scanning a specific file, like `occ files:scan --path`?

**Comment by steffenclemens** [comment]
@Ornanovitch `occ groupfolders:scan -h` shows you all options with description.

**Comment by Ornanovitch** [comment]
I just tested your patch and it seems something's wrong :/

"/CLASSEUR AFFAIRES 1999" is at the root of the groupfolder 12, and it contains only 14 folders and 1 file:

```log
cloud :: /var/www/nextcloud » sudo -u www-data php --define apc.enable_cli=1 occ groupfolders:scan 12 --path="/CLASSEUR AFFAIRES 1999" --shallow
+-----------+---------+-------+--------------+
| Folder Id | Folders | Files | Elapsed time |
+-----------+---------+-------+--------------+
| 12        | 98899   | 11343 | 00:01:43     |
+-----------+---------+-------+--------------+
```

**Comment by Ornanovitch** [comment]
Just tested your patch, works very well but there is something strange about the files/folder count:

`/CLASSEUR AFFAIRES 1999` contains 18 folders and 1 file:

```log
» sudo -u www-data php occ groupfolders:scan 12 --path="/CLASSEUR AFFAIRES 1999" --shallow
+-----------+---------+-------+--------------+
| Folder Id | Folders | Files | Elapsed time |
+-----------+---------+-------+--------------+
| 12        | 1       | 1     | 00:00:00     |
+-----------+---------+-------+--------------+
```

Nothing's wrong on the UI side

**Comment by Ornanovitch** [comment]
Note that it works with a single file :partying_face:

**Comment by Ornanovitch** [comment]
@icewind1991 any chance to merge this awesome feature?

**Comment by skjnldsv** [comment]
@Ornanovitch rebased and pinged more devs for this :)
cc @icewind1991, please address your review again :bow:

**Review by skjnldsv — APPROVED** [review]

**Review by provokateurin — APPROVED** [review]

**Comment by provokateurin** [comment]
Sorry, we don't backport features.

**Comment by Ornanovitch** [comment]
I understand but does it mean this feature won't be able before NC 31? That would be unexpected, because it was ready before the milestone even exist.. :/

**Comment by provokateurin** [comment]
It was just merged too later after the branch off.

**Pull request merged**